### PR TITLE
[WIP] Piezo Sensitivity and Site symmetries packages added

### PR DESCRIPTION
**Pull request by dongsenfo**
WIP

## Summary

Added piezo_sensitivity and site_symmetry packages under analysis. Site_symmetries allow for users to get the point symmetries of atomic sites and symmetry operations relating sites. Piezo sensitivity allows for random generation of piezoelectric components and the calculation of the piezoelectric tensor from components

## TODO (if any)

Needs to be unit tested and style formatted. Not quite sure if this is the appropriate place for the packages but these are still heavily work in progress.

### CODE REVIEW

**Comment by shyuep** [comment]
Thanks for the contributions. Yes, you can add this to pymatgen.analysis. However, I would ask that you document the methods. Site symmetries should be put in the pymatgen.symmetry.analyzer module.

**Comment by shyamd** [comment]
Are the site symmetry routines not uses in the PIezo sensitivity analysis?

Things to do:
- Write a Born Effective Charge (BEC), Internal Strain Tensor (IST), and Force Constant Matrix (FCM) tensor class; @shyuep Can we start collecting tensor quantities in a root level tensors module?
- Add in the construction of the Piezo Tensor from BEC,IST, and FCM into the PiezoTensor Class
- Add documentation
- Make variable names more explicit so we don't have to guess what they are
- We'll work on some algo clean up

**Comment by dongsenfo** [comment]
Added documentation and classes for the separate tensors

File: pymatgen/analysis/piezo_sensitivity.py [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+        Create an BornEffectiveChargeTensor object.  The constructor throws 
+        an error if the input_matrix argument does not satisfy the acoustic 
+        sum rule. Note that the constructor uses __new__ rather than __init__ 
+        according to the standard method of subclassing numpy ndarrays.
+
+        Args:
+            input_matrix (Nx3x3 array-like): the Nx3x63 array-like
+                representing the born effective charge tensor
+        """
+        obj = super(PiezoTensor, cls).__new__(cls, input_array, check_rank=4)
```
Comment (shyamd) [inline pymatgen/analysis/piezo_sensitivity.py]: Make sure you properly subclassing Tensor in the new statements

File: pymatgen/analysis/piezo_sensitivity.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+        if not (np.sum(obj)) < tol).all():
+            warnings.warn("Input born effective charge tensor does "
+                          "not satisfy acoustic sum rule")
```
Comment (shyamd) [inline pymatgen/analysis/piezo_sensitivity.py]: I believe for BEC it's actually charge neutrality not acoustic sum here.

File: pymatgen/analysis/piezo_sensitivity.py [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+        Args:
+            input_matrix (NxNx3x3 array-like): the NxNx3x3 array-like
+                representing the force constant matrix
+        """
+        obj = super(PiezoTensor, cls).__new__(cls, input_array, check_rank=4)
+
+        return obj.view(cls)
+
+
+def get_rand_BEC(struc, eqop, symops, max_charge=1):
```
Comment (shyamd) [inline pymatgen/analysis/piezo_sensitivity.py]: These functions should be methods of the corresponding Tensor class. Symops should probably default to None, and if so then you use SpaceGroupAnalyzer to find the symops. If not, you use the given symops.

File: pymatgen/analysis/piezo_sensitivity.py [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+        temp_tensor = Tensor(np.random.rand(3,3)-0.5)         
+        temp_tensor = sum([temp_tensor.transform(symm_op) for symm_op in symops[i]]) \
+            /len(symops[i])
+            
+        BEC[i] = np.array(temp_tensor)
+
+    count = 0
+
+    
+#     Enforce Acoustic Sum
```
Comment (shyamd) [inline pymatgen/analysis/piezo_sensitivity.py]: These should be submethods of the appropriate class as well. Then you can just call it. This makes the code a bit cleaner to read.

File: pymatgen/analysis/piezo_sensitivity.py [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+
+        add[i] = sum([Tensor(add[i]).transform(symm_op) for symm_op in symops[i]]) / len(symops[i])
+    
+    BEC = BEC - add
+    
+    disp_charge = np.einsum("ijk->jk",BEC)/l
+ 
+    return BornEffectiveChargeTensor(BEC*max_charge)
+
+def get_rand_IST(struc, eqop, symops, max_force=1):
```
Comment (shyamd) [inline pymatgen/analysis/piezo_sensitivity.py]: Same as before, make this a submethod of IST

File: pymatgen/analysis/piezo_sensitivity.py [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+        temp_tensor = sum([temp_tensor.transform(symm_op) for symm_op in symops[i]]) \
+            /len(symops[i])
+ 
+        IST[i] = np.array(temp_tensor)
+    count = 0
+
+    return InternalStrainTensor(IST*max_force)
+
+
+def get_FCM_symmetry(fcm, operations, sharedops, max_force = 1):
```
Comment (shyamd) [inline pymatgen/analysis/piezo_sensitivity.py]: same as before, make this a submethod of FCM.

File: pymatgen/analysis/piezo_sensitivity.py [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+        force constant matrix representing the force constant matrix
+    """
+    numsites = len(struc.sites)
+    structure = phonopy.get_phonopy_structure(struc)
+    operations = get_fcm_operations(struc, fcm)
+    pnstruc = pn.Phonopy(structure, np.eye(3), np.eye(3))
+    symops = ss.get_site_symmetries(struc)
+    sharedops = ss.get_shared_symmetry_operations(strucs, symops)
+
+    fcm  = get_FCM2(struc, sharedops, operations, force)
```
Comment (shyamd) [inline pymatgen/analysis/piezo_sensitivity.py]: get_FCM2 doesn't seem to be defined anywhere?

Comment (shyamd) [reply pymatgen/analysis/piezo_sensitivity.py]: Also, use descriptive names for methods.

File: pymatgen/analysis/piezo_sensitivity.py [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+                    new = ops[j].transform_tensor(dyn[relations[i][2]][relations[i][3]])
+                    if np.allclose(new, dyn[relations[i][1]][relations[i][0]], atol = 1e-05):
+                        transpose.append(ops[j])
+                operations[i].append(transpose)
+                break
+        
+    return operations
+
+                    
+def rand_piezo(struc, fcm, anumiter = 10):
```
Comment (shyamd) [inline pymatgen/analysis/piezo_sensitivity.py]: This should probably a submethod of piezo tensor, but for now, we can leave this here.

**Comment by shyamd** [comment]
Ok, fix the conflicts from the main branch. Then unit tests for all the functions. That will allow you to see what is still broken. Also expand out your abbreviations. It's better for public code to read almost like prose.

**Comment by shyamd** [comment]
Ooops, I don't think this was ready to be merged yet. @dongsenfo  Can you make a new PR that addresses the outlined issues?

**Pull request merged**